### PR TITLE
CAS client API refactor; Shard and CAS object format changes to use chunk indices

### PR DESCRIPTION
**Pull request by seanses**
- Delete grpc cas and shard code.
- Rewrite CAS client api as there will be no API to get cas block by hash.
- Delete CachingClient and StagingClient (need to rewrite them in the future).
- Use chunk indices in `FileDataSequenceEntry` and cas object format.
- Misc code clean up and reorganization.

Related server-side PR:

- https://github.com/huggingface-internal/xetcas/pull/46
- https://github.com/huggingface-internal/xetcas/pull/53

### CODE REVIEW

File: cas_object/src/cas_chunk_format.rs [hunk 1]
```diff
@@ -145,14 +145,23 @@ pub fn deserialize_chunk_to_writer<R: Read, W: Write>(
...
-        CompressionScheme::None => writer.write_all(&compressed_buf)?,
+    let uncompressed_len = match header.get_compression_scheme() {
+        CompressionScheme::None => {
+            writer.write_all(&compressed_buf)?;
+            compressed_buf.len() as u32
+        }
         CompressionScheme::LZ4 => {
             let mut dec = FrameDecoder::new(Cursor::new(compressed_buf));
-            copy(&mut dec, writer)?;
+            copy(&mut dec, writer)? as u32
```
Comment (assafvayner) [inline cas_object/src/cas_chunk_format.rs:155]: IIRC this will return the compressed len!

Comment (assafvayner) [reply cas_object/src/cas_chunk_format.rs:155]: Actually, not sure anymore.

File: cas_client/src/remote_client.rs [hunk 2]
```diff
@@ -280,14 +199,13 @@ async fn get_one(term: &CASReconstructionTerm) -> Result<Bytes> {
         .bytes()
         .await
         .map_err(CasClientError::ReqwestError)?;
+    if xorb_bytes.len() as u32 != term.url_range.end - term.url_range.start {
+        error!("got back a smaller range than requested");
```
Comment (rajatarya) [inline cas_client/src/remote_client.rs:254]: Should this return an Err instead of just logging?

File: shard_client/src/http_shard_client.rs [hunk 3]
```diff
@@ -182,17 +195,66 @@ impl ShardDedupProber<ShardClientError> for HttpShardClient {
...
 
         let response_body = response.bytes().await?;
-        let response_info: QueryChunkResponse = serde_json::from_reader(response_body.reader())?;
+        let mut reader = response_body.reader();
+
+        let mut downloaded_shards = vec![];
+
+        let mut write_join_set = JoinSet::<Result<()>>::new();
+
+        // Return in the format of
```
Comment (rajatarya) [inline shard_client/src/http_shard_client.rs:204]: What concerns me with this loop is that the server can "DOS" the client by just continuously sending 'bytes'. Instead, could this protocol start with num_shards: usize and then for each shard: [key, shard] (like it is here)?

Comment (rajatarya) [reply shard_client/src/http_shard_client.rs:204]: Also, what happens if the `read_u32` call fails for another reason (ex. 'broken pipe') - shouldn't that error propagate up to trigger a retry?

File: utils/src/singleflight.rs
Comment (rajatarya) [inline utils/src/singleflight.rs:1]: Should we delete this file?

File: cas_object/src/cas_chunk_format.rs (hunk 1 above)
Comment (seanses) [reply cas_object/src/cas_chunk_format.rs:155]: The number of bytes read and that written should equal: `copy` doesn't do any magic, it just does byte-by-byte copy. It's the `impl Read for FrameDecoder` that generates more bytes than what's read from `compressed_buf`.

File: utils/src/singleflight.rs
Comment (seanses) [reply utils/src/singleflight.rs:1]: Maybe not so early. We may need it again.

File: cas_client/src/remote_client.rs (hunk 2 above)
Comment (seanses) [reply cas_client/src/remote_client.rs:254]: Fixed.

File: shard_client/src/http_shard_client.rs (hunk 3 above)
Comment (seanses) [reply shard_client/src/http_shard_client.rs:204]: > What concerns me with this loop is that the server can "DOS" the client by just continuously sending 'bytes'. Instead, could this protocol start with num_shards: usize and then for each shard: [key, shard] (like it is here)?

Rather, maybe the client can define some threshold that "I have had enough dedup info downloaded", and stop reading from the socket.

Comment (seanses) [reply shard_client/src/http_shard_client.rs:204]: > Also, what happens if the `read_u32` call fails for another reason (ex. 'broken pipe') - shouldn't that error propagate up to trigger a retry?

Actually we didn't implement streaming: the `response.bytes()` above already collected all bytes so there's no pipe here.

Comment (seanses) [reply shard_client/src/http_shard_client.rs:204]: I checked multiple places and looks like we never retry on corrupted response. I guess that will be severe bug and it's not worthy to retry?

**Pull request merged**